ksql-engine/src/main/java/io/confluent/ksql/codegen/SqlToJavaVisitor.java

```diff
@@ -633,15 +633,30 @@ private String buildSupplierCode(final String typeString, final String code) {
           SchemaUtil.getJavaType(internalSchema).getCanonicalName();
       switch (internalSchema.type()) {
         case ARRAY:
-          return new Pair<>(
-              String.format("((%s) ((%s)%s).get((int)(%s)))",
-                  SchemaUtil.getJavaType(internalSchema.valueSchema()).getSimpleName(),
-                  internalSchemaJavaType,
-                  process(node.getBase(), context).getLeft(),
-                  process(node.getIndex(), context).getLeft()
-              ),
-              internalSchema.valueSchema()
-          );
+          final String listName = process(node.getBase(), context).getLeft();
+          if (node.getIndex().toString().startsWith("-")) {
```

Comment:
would you mind refactoring this a little so that the index part is calculated in the if/else, and the rest is in a common part?

e.g.

```java
final String suppliedIdx = process(node.getIndex(), context).getLeft();
final String trueIdx = node.getIndex().toString().startsWith("-")
   ? String.format("(%s)%s).size()%s", internalSchemaJavaType, listName, suppliedIdx)
   : suppliedIdx;

return new Pair<>(
  String.format(((%s) ((%s)%s).get((int)(%s)", 
       SchemaUtil.getJavaType(internalSchema.valueSchema()).getSimpleName(),
      internalSchemaJavaType,
      listName,
      trueIdx),
  internalSchema.valueSchema()
);
```

Reply:
@big-andy-coates https://github.com/confluentinc/ksql/pull/3004/commits/ef7a82a0c8dcd51bbbc37a602cb74f3f41596ae2